File: StackStackStack/StackStackStack/Capture/CaptureOrientation.swift
import UIKit

/// Maps the physical device orientation at shutter time to the clockwise quarter-turns needed to make
/// the back camera's native-landscape stacked result upright. The constants are validated on a
/// physical device (the back wide-camera buffer is landscape-native); see the plan's device-verify step.
enum CaptureOrientation {
    static func quarterTurns(for orientation: UIDeviceOrientation) -> Int {
        switch orientation {
        case .portrait:           return 1
        case .portraitUpsideDown: return 3
        case .landscapeLeft:      return 2
        case .landscapeRight:     return 0
        default:                  return 1   // faceUp/faceDown/unknown → assume portrait
        }
    }
}

File: android/app/src/main/kotlin/com/jdpneto/stackstackstack/CaptureOrientation.kt
package com.jdpneto.stackstackstack

import android.view.Surface

/**
 * Maps the physical device orientation at shutter time to the clockwise quarter-turns needed
 * to make the back camera's native-landscape stacked result upright.
 *
 * Device-verified: portrait shows upright, and the two landscape cases are *swapped* relative
 * to the naive guess — [Surface.ROTATION_90] / [Surface.ROTATION_270] correspond to the
 * opposite video orientations (the classic landscape inversion), so a landscape shot otherwise
 * saves upside-down.
 *
 * Android uses [Surface.ROTATION_*] constants (0=portrait, 90=landscape-left, etc.).
 * Mirrors iOS [CaptureOrientation.quarterTurns(for:)] with the same mapping table.
 */
object CaptureOrientation {
    /**
     * Map a [Surface] rotation constant to the clockwise quarter-turns needed to upright the
     * result image.
     *
     * @param displayRotation One of [Surface.ROTATION_0], [Surface.ROTATION_90],
     *   [Surface.ROTATION_180], [Surface.ROTATION_270].
     */
    fun quarterTurns(displayRotation: Int): Int = when (displayRotation) {
        Surface.ROTATION_0   -> 1   // portrait
        Surface.ROTATION_180 -> 3   // portrait upside-down
        Surface.ROTATION_90  -> 0   // landscape (left home button) — maps to ROTATION_90
        Surface.ROTATION_270 -> 2   // landscape (right home button) — maps to ROTATION_270
        else                 -> 1   // unknown → assume portrait
    }
}
